src/acl/agent.ts

```diff
@@ -369,52 +372,36 @@ function duplicateAclRule(sourceRule: unstable_AclRule): unstable_AclRule {
  * @param resourceIri The Resource to which the Rule should no longer apply for the given Agent.
  * @returns A tuple with the original ACL Rule sans the given Agent, and a new ACL Rule for the given Agent for the remaining Resources, respectively.
  */
-function removeAgentFromResourceRule(
+function removeAgentFromRule(
   rule: unstable_AclRule,
   agent: WebId,
-  resourceIri: IriString
+  resourceIri: IriString,
+  ruleType: "resource" | "default"
 ): [unstable_AclRule, unstable_AclRule] {
   // The existing rule will keep applying to Agents other than the given one:
   const ruleWithoutAgent = removeIri(rule, acl.agent, agent);
-  // The new rule will...
-  let ruleForOtherTargets = duplicateAclRule(rule);
-  // ...*only* apply to the given Agent (because the existing Rule covers the others)...
+  let ruleForOtherTargets: unstable_AclRule;
+  if (getIriAll(rule, acl.agent).includes(agent)) {
+    // The agent already had some access in the rule, so duplicate it...
+    ruleForOtherTargets = duplicateAclRule(rule);
+    // ...but remove access to the original Resource:
+    ruleForOtherTargets = removeIri(
+      ruleForOtherTargets,
+      ruleType === "resource" ? acl.accessTo : acl.default,
+      resourceIri
+    );
+  } else {
+    // If the agent had no access, the new ACL rule is empty
+    ruleForOtherTargets = intialiseAclRule({
+      read: false,
+      append: false,
+      write: false,
+      control: false,
+    });
+  }
```

Comment:
Ah, clever - I was thinking about returning `null` and then checking for that on the call-site, but given that empty rules are removed afterwards, this should fix that.

That said, I'd move this around a bit to make it an early exit condition:

```suggestion
  if(!getIriAll(rule, acl.agent).includes(agent)) {
    // If the agent had no access, We don't need to split this rule up:
    const emptyRule = intialiseAclRule({
      read: false,
      append: false,
      write: false,
      control: false,
    });
    return [rule, emptyRule];
  }
  // The existing rule will keep applying to Agents other than the given one: 
  const ruleWithoutAgent = removeIri(rule, acl.agent, agent);
  let ruleForOtherTargets: unstable_AclRule;
  // The agent already had some access in the rule, so duplicate it...
  ruleForOtherTargets = duplicateAclRule(rule);
  // ...but remove access to the original Resource:
  ruleForOtherTargets = removeIri(
    ruleForOtherTargets,
    ruleType === "resource" ? acl.accessTo : acl.default,
    resourceIri
  );
```

Reply:
Resolved in a9fd918d3f3062c014dd3729448ff3e52a8b691d